xline-client/src/clients/watch.rs

```diff
@@ -26,4 +29,267 @@ impl WatchClient {
             )),
         }
     }
+
+    /// Watches for events happening or that have happened. Both input and output
+    /// are streams; the input stream is for creating and canceling watcher and the output
+    /// stream sends events. The entire event history can be watched starting from the
+    /// last compaction revision.
+    ///
+    /// # Errors
+    ///
+    /// If the RPC client fails to send request
+    ///
+    /// # Panics
+    ///
+    /// If the RPC server returns the wrong value
+    #[inline]
+    pub async fn watch(
+        &mut self,
+        request: WatchRequest,
+    ) -> Result<(Watcher, tonic::Streaming<WatchResponse>)> {
+        let (mut request_sender, request_receiver) = channel::<xlineapi::WatchRequest>(100);
+
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(request.into())),
+        };
+
+        request_sender
+            .try_send(request)
+            .map_err(|e| ClientError::WatchError(e.to_string()))?;
+
+        let mut response_stream = self.inner.watch(request_receiver).await?.into_inner();
+
+        let watch_id = match response_stream.message().await? {
+            Some(resp) => {
+                assert!(resp.created, "not a create watch response");
+                resp.watch_id
+            }
+            None => {
+                return Err(ClientError::WatchError(String::from(
+                    "failed to create watch",
+                )));
+            }
+        };
+
+        Ok((Watcher::new(watch_id, request_sender), response_stream))
+    }
+}
+
+/// The watching handle.
+#[derive(Debug)]
+pub struct Watcher {
+    /// Id of the watcher
+    watch_id: i64,
+    /// The channel sender
+    sender: Sender<xlineapi::WatchRequest>,
+}
+
+impl Watcher {
+    /// Creates a new `Watcher`.
+    #[inline]
+    const fn new(watch_id: i64, sender: Sender<xlineapi::WatchRequest>) -> Self {
+        Self { watch_id, sender }
+    }
+
+    /// The ID of the watcher.
+    #[inline]
+    #[must_use]
+    pub const fn watch_id(&self) -> i64 {
+        self.watch_id
+    }
+
+    /// Watches for events happening or that have happened.
+    ///
+    /// # Errors
+    ///
+    /// If sender fails to send to channel
+    #[inline]
+    pub fn watch(&mut self, request: WatchRequest) -> Result<()> {
+        let request = xlineapi::WatchRequest {
+            request_union: Some(RequestUnion::CreateRequest(request.into())),
```

Comment:
What are the differences between `WatchClient::watch` and `Watcher::watch`? Both of them send the same `CreateRequest` via `self.sender`. Shall we create twice?

Reply:
The Watcher is used to add new watch request or cancel existing request interactively, while WatchClient::watch can only watch on a const range of keys.